File: 01-Programming-Languages/TypeScript/Data-Structures/LinkedList/linked_list.ts
class ListNode {
    value: number;
    next: ListNode | null;

    constructor(value: number) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    head: ListNode | null;

    constructor() {
        this.head = null;
    }

    append(value: number): void {
        const newNode = new ListNode(value);

        if (!this.head) {
            this.head = newNode;
            return;
        }

        let current = this.head;
        while (current.next) {
            current = current.next;
        }
        current.next = newNode;
    }

    prepend(value: number): void {
        const newNode = new ListNode(value);
        newNode.next = this.head;
        this.head = newNode;
    }

    delete(value: number): void {
        if (!this.head) return;

        if (this.head.value === value) {
            this.head = this.head.next;
            return;
        }

        let current = this.head;
        while (current.next && current.next.value !== value) {
            current = current.next;
        }

        if (current.next) {
            current.next = current.next.next;
        }
    }

    display(): void {
        let current = this.head;
        while (current) {
            process.stdout.write(`${current.value} -> `);
            current = current.next;
        }
        console.log("null");
    }
}

const list = new LinkedList();
list.append(1);
list.append(2);
list.append(3);
list.prepend(0);

console.log("List:");
list.display();

list.delete(2);
console.log("After deleting 2:");
list.display();

File: 01-Programming-Languages/Dart/Data-Structures/LinkedList/linked_list.dart
class Node {
  int value;
  Node? next;

  Node(this.value) : next = null;
}

class LinkedList {
  Node? head;

  void append(int value) {
    final newNode = Node(value);
    if (head == null) {
      head = newNode;
      return;
    }

    var current = head!;
    while (current.next != null) {
      current = current.next!;
    }
    current.next = newNode;
  }

  void prepend(int value) {
    final newNode = Node(value);
    newNode.next = head;
    head = newNode;
  }

  void delete(int value) {
    if (head == null) return;

    if (head!.value == value) {
      head = head!.next;
      return;
    }

    var current = head!;
    while (current.next != null && current.next!.value != value) {
      current = current.next!;
    }
    if (current.next != null) {
      current.next = current.next!.next;
    }
  }

  void display() {
    var current = head;
    while (current != null) {
      stdout.write('${current.value} -> ');
      current = current.next;
    }
    print('null');
  }
}

void main() {
  final list = LinkedList();
  list.append(1);
  list.append(2);
  list.append(3);
  list.prepend(0);

  print('List:');
  list.display();

  list.delete(2);
  print('After deleting 2:');
  list.display();
}